core/chains/evm/gas/rollups/op_l1_oracle.go

```diff
@@ -0,0 +1,490 @@
+
+	priceReader daPriceReader
+
+	chInitialised chan struct{}
+	chStop        services.StopChan
+	chDone        chan struct{}
+}
+
+type opStackGasPriceReader struct {
+	client l1OracleClient
+	logger logger.SugaredLogger
+
+	oracleAddress      common.Address
+	isEcotoneMethodAbi abi.ABI
+
+	l1BaseFeeCalldata    []byte
+	isEcotoneCalldata    []byte
+	getL1GasUsedCalldata []byte
+	getL1FeeCalldata     []byte
+
+	isEcotone        bool
+	isEcotoneCheckTs int64
+}
+
+const (
+	// OPStackGasOracle_l1BaseFee fetches the l1 base fee set in the OP Stack GasPriceOracle contract
+	OPStackGasOracle_l1BaseFee = "l1BaseFee"
+
+	// OPStackGasOracle_isEcotone fetches if the OP Stack GasPriceOracle contract has upgraded to Ecotone
+	OPStackGasOracle_isEcotone = "isEcotone"
+
+	// OPStackGasOracle_getL1GasUsed fetches the l1 gas used for given tx bytes
+	OPStackGasOracle_getL1GasUsed = "getL1GasUsed"
+
+	// OPStackGasOracle_getL1Fee fetches the l1 fee for given tx bytes
+	OPStackGasOracle_getL1Fee = "getL1Fee"
+
+	// OPStackGasOracle_isEcotonePollingPeriod is the interval to poll if chain has upgraded to Ecotone
+	// Set to poll every 4 hours
+	OPStackGasOracle_isEcotonePollingPeriod = 14400
+	// OPGasOracleAddress is the address of the precompiled contract that exists on OP stack chain.
+	// This is the case for Optimism and Base.
+	OPGasOracleAddress = "0x420000000000000000000000000000000000000F"
+	// OPGasOracle_l1BaseFee is a hex encoded call to:
+	// `function l1BaseFee() external view returns (uint256);`
+	OPGasOracle_l1BaseFee = "l1BaseFee"
+	// OPGasOracle_getL1Fee is a hex encoded call to:
+	// `function getL1Fee(bytes) external view returns (uint256);`
+	OPGasOracle_getL1Fee = "getL1Fee"
+
+	// GasOracleAddress is the address of the precompiled contract that exists on Kroma chain.
+	// This is the case for Kroma.
+	KromaGasOracleAddress = "0x4200000000000000000000000000000000000005"
+	// GasOracle_l1BaseFee is the a hex encoded call to:
+	// `function l1BaseFee() external view returns (uint256);`
+	KromaGasOracle_l1BaseFee = "l1BaseFee"
+)
+
+func NewOpStackL1GasOracle(lggr logger.Logger, ethClient l1OracleClient, chainType config.ChainType) L1Oracle {
+	var precompileAddress string
+	switch chainType {
+	case config.ChainOptimismBedrock:
+		precompileAddress = OPGasOracleAddress
+	case config.ChainKroma:
+		precompileAddress = KromaGasOracleAddress
+	default:
+		panic(fmt.Sprintf("Received unspported chaintype %s", chainType))
+	}
+	priceReader := newOPPriceReader(lggr, ethClient, chainType, precompileAddress)
+	return newOpStackL1GasOracle(lggr, ethClient, priceReader, chainType)
+}
+
+func newOpStackL1GasOracle(lggr logger.Logger, ethClient l1OracleClient, priceReader daPriceReader, chainType config.ChainType) L1Oracle {
+	var l1GasPriceAddress, gasPriceMethod, l1GasCostAddress, gasCostMethod string
+	var l1GasPriceMethodAbi, l1GasCostMethodAbi abi.ABI
+	var gasPriceErr, gasCostErr error
+
+	l1GasPriceAddress = OPGasOracleAddress
```

Comment:
Think we'll also want to use the `precompileAddress` from `NewOpStackL1GasOracle` in this method too so we call the right contract. Otherwise from Kroma we'd be calling the wrong address.